File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.go
func subarraysDivByK(nums []int, k int) int {
    sum := 0
    res := 0
    mp := make(map[int]int)
    mp[0] = 1

    for _, val := range nums {
        sum = sum + val
        rem := sum % k
        if rem < 0 {
            rem = rem + k
        }
        res = res + mp[rem]
        mp[rem]++
    }
    return res
}

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int sum = 0;
        int res = 0;
        unordered_map<int, int> mp;
        mp[0] = 1;

        for (int num : nums) {
            sum = sum + num;
            int rem = sum % k;
            if (rem < 0) rem = rem + k;
            res = res + mp[rem];
            mp[rem]++;
        }
        
        return res;
    }
};